lib/multiplexer/multiplexer.go

```diff
@@ -400,20 +400,24 @@ func isHTTP(in []byte) bool {
 		[]byte("PATCH"),
 	}
 	for _, verb := range methods {
-		// we only get 3 bytes, so can only compare the first 3 bytes of each verb
-		if bytes.HasPrefix(verb, in[:3]) {
+		if bytes.HasPrefix(verb, in) {
 			return true
 		}
 	}
 	return false
 }
 
+// detectProto tries to determine the network protocol used from the first
+// 8 bytes of a connection.
 func detectProto(in []byte) (Protocol, error) {
+	if len(in) < 8 {
+		return ProtoUnknown, trace.BadParameter("expected at least 8 bytes, got %d", len(in))
+	}
```

Comment:
Is there a reason we want to enforce this? SSH/TLS/Proxy all need fewer than 8 bytes, right?

Reply:
Proxy v1 needs 5 bytes, Proxy v2 needs 12, and HTTP methods need max 7 or 8 I believe.

I added this check so that the rest of the function can safely assume `in` is at least 8 bytes in length when doing protocol checks. All current callers of `detectProto` always pass in 8 bytes, so will just be helpful for avoiding subtle errors in new code.

Reply:
What if we add a protocol that sends less than 8 bytes in the first packet? It's theoretical of course, just wondering if we're rather opening a way for more problems later.

Reply:
That's an interesting point... The main reason I created this PR was I didn't think it made sense to try and match the first few bytes of a connection with the magic bytes of a protocol when the bytes given to match are shorter than some of the magic byte sequences. We can't be sure that magic bytes match unless we match against all magic bytes. Maybe I'm thinking of this the wrong way..